akd_client/src/tests.rs

```diff
@@ -194,6 +269,120 @@ async fn test_simple_lookup() -> Result<(), AkdError> {
     Ok(())
 }
 
+#[tokio::test]
+async fn test_history_proof_multiple_epochs() -> Result<(), AkdError> {
+    let db = InMemoryDb::new();
+    let vrf = HardCodedAkdVRF {};
+    let akd = Directory::new::<Hash>(&db, &vrf, false).await?;
+    let vrf_pk = akd.get_public_key().await.unwrap();
+    let key = AkdLabel("label".to_string());
+    let key_bytes = key.0.as_bytes().to_vec();
+    const EPOCHS: usize = 10;
+
+    // publishes key versions in multiple epochs
+    for epoch in 1..=EPOCHS {
+        let data = vec![(key.clone(), AkdValue(format!("value{}", epoch)))];
+        akd.publish::<Hash>(data, true).await?;
+    }
+
+    // retrieves and verifies history proofs for the key
+    let proof = akd.key_history::<Hash>(&key).await?;
+    let internal_proof = convert_history_proof::<Hash>(&proof);
+    let (mut root_hashes, previous_root_hashes) =
+        akd::directory::get_key_history_hashes::<_, Hash, HardCodedAkdVRF>(&akd, &proof)
+            .await
+            .unwrap();
+
+    // verifies num of root hashes created
+    assert_eq!(root_hashes.len(), EPOCHS);
+
+    // verifies both traditional and lean history verification passes
+    {
+        let akd_result = akd::client::key_history_verify::<Hash>(
+            &vrf_pk,
+            root_hashes.clone(),
+            previous_root_hashes.clone(),
+            key.clone(),
+            proof.clone(),
+        );
+        let lean_result = crate::verify::key_history_verify(
+            &vrf_pk.to_bytes(),
+            to_digest_vec::<Hash>(root_hashes.clone()),
+            to_digest_vec_opt::<Hash>(previous_root_hashes.clone()),
+            key_bytes.clone(),
+            internal_proof.clone(),
+        );
+        assert!(matches!(akd_result, Ok(())), "{:?}", akd_result);
```

Comment:
It does not print akd_result for me without custom message. I've just retried to confirm that.

Reply:
+1 on contexts! We have an issue for it #157.

Lack of custom messages might be due to [this](https://github.com/novifinancial/akd/issues/157#issuecomment-1058348055).

Reply:
Ah because we're using `matches!`, the `panic!` statement will only state that `matches!` resulted in `false`. The custom message is needed to dump the value unfortunately.